### PR TITLE
KAFKA-3950: kafka mirror maker tool is not respecting whitelist option

### PR DESCRIPTION
**Pull request by omkreddy**

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java [hunk 1]
```diff
@@ -146,11 +146,16 @@ private void addMetadataListener() {
...
-                if (!cluster.unauthorizedTopics().isEmpty())
-                    throw new TopicAuthorizationException(new HashSet<>(cluster.unauthorizedTopics()));
 
                 if (subscriptions.hasPatternSubscription()) {
+
+                    for (String topic : cluster.unauthorizedTopics()) {
+                        if (subscriptions.getSubscribedPattern().matcher(topic).matches())
+                            throw new TopicAuthorizationException(topic);
+                        else
+                            log.debug("Ignoring unauthorized topic {}, which not matches with subscribed pattern", topic);
```
Comment (SinghAsDev) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java]: Typo: "which not matches with"

Comment (SinghAsDev) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java]: Not sure if we want to provide information on a topic on which user does not have permissions.

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala [hunk 2]
```diff
@@ -48,6 +50,7 @@ class AuthorizerIntegrationTest extends KafkaServerTestHarness {
   val correlationId = 0
   val clientId = "client-Id"
   val tp = new TopicPartition(topic, part)
+  val topicpattren = "topic*"
```
Comment (SinghAsDev) [inline core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala]: maybe use `topicPattern` instead.

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala [hunk 3]
```diff
@@ -415,6 +418,18 @@ class AuthorizerIntegrationTest extends KafkaServerTestHarness {
   }
 
   @Test
+  def testPattrenSubscriptionWithTopicAndGroupRead() {
```
Comment (SinghAsDev) [inline core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala]: Adding a negative test will probably be a good idea as well, if it is not there already.

**Comment by SinghAsDev** [comment]
Left a few comments, otherwise LGTM.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java (hunk 1 above)
Comment (harshach) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java]: don't think we need to log here . If a topic is not matching to the pattern than why bother logging here.

**Comment by omkreddy** [comment]
@SinghAsDev @harshach Thanks for the reviews. addressed review comments

**Comment by rajinisivaram** [comment]
@omkreddy The changes in `ConsumerCoordinator` are inconsistent with the proposal in https://issues.apache.org/jira/browse/KAFKA-3396 to address the security issue in the current code that returns unauthorized topics to clients. An option would be to ignore unauthorized topics that match patterns and only subscribe to authorized topics that match the pattern without throwing an exception.

**Comment by omkreddy** [comment]
@rajinisivaram  Thanks for the details.  My concern is ignoring unauthorized topics will make debugging harder for authorization related issues. Since this is a critical issue, may be we can fix now and we can revisit after KAFKA-3396 and KIP-4.

**Comment by harshach** [comment]
@rajinisivaram  agree with @omkreddy here. Its not just debugging but a unexpected behavior where users might be thinking they subscribed to all the topics and internally we are skipping them due to authorization issues.

**Comment by rajinisivaram** [comment]
@omkreddy @harshach Agree that it is ok to apply this fix now and revisit in KAFKA-3396 since this only returns the same information as the code already does.

**Comment by ijuma** [comment]
@harshach, returning all topic names to the client even if the client doesn't have `Describe Topic` permission is an information leak and we definitely need to fix that (as per KAFKA-3396). This can lead to unexpected behaviour in this case (as I said in the JIRA for this PR), but I don't see a way around that.

~~In any case, the suggested change still makes sense once KAFKA-3396 lands as even if we don't return the unauthorized topics that the user doesn't have `Describe Topic` permission for, there is still the case where the user has `Describe Topic` permission, but no `Create` permission for the auto-topic creation case.~~ Edit: Auto-creation of topics doesn't happen for pattern subscriptions, so what I had said about that doesn't make sense. Given that, it's less clear if this change makes sense for trunk. Maybe for `0.10.0` though.

cc @hachikuji

**Comment by harshach** [comment]
@ijuma I am +1 on merging this. current behavior doesn't sound right.

**Comment by ijuma** [comment]
@harshach, I am suggesting that the approach in KAFKA-3396 is the right one for trunk. The broker won't populate unauthorized topics for topics that the user doesn't have Describe permissions for. With that change, this change doesn't seem to make sense.

The change here could be OK for the 0.10.0 branch though as it's more conservative.

**Comment by harshach** [comment]
@ijuma pinging. Lets include this in 10.0.1.0 release.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java [hunk 4]
```diff
@@ -146,11 +146,14 @@ private void addMetadataListener() {
...
             public void onMetadataUpdate(Cluster cluster) {
-                // if we encounter any unauthorized topics, raise an exception to the user
-                if (!cluster.unauthorizedTopics().isEmpty())
-                    throw new TopicAuthorizationException(new HashSet<>(cluster.unauthorizedTopics()));
 
                 if (subscriptions.hasPatternSubscription()) {
+
+                    for (String topic : cluster.unauthorizedTopics()) {
+                        if (subscriptions.getSubscribedPattern().matcher(topic).matches())
+                            throw new TopicAuthorizationException(topic);
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java]: Maybe we may as well collect all the matching unauthorized topics before raising?

**Comment by hachikuji** [comment]
I agree with @ijuma that this makes less sense in trunk, but the impact of the bug is pretty bad, so if we can't have the patch for KAFKA-3396 in 0.10.0, we should probably include this.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java (hunk 4 above)
Comment (omkreddy) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java]: Thanks @hachikuji. Updated PR.

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala [hunk 5]
```diff
@@ -415,6 +418,34 @@ class AuthorizerIntegrationTest extends KafkaServerTestHarness {
   }
 
   @Test
+  def testPatternSubscriptionWithNoTopicAccess() {
```
Comment (hachikuji) [inline core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:411]: This verifies that unauthorized topics matching the regex will cause the authorization exception to be thrown, but should we also have a test case which verifies that non-matching unauthorized topics do not cause the exception to be thrown?

**Comment by ijuma** [comment]
The PR should target the 0.10.0 branch if we agree that the aim is to merge this there (and not trunk).

**Comment by omkreddy** [comment]
@ijuma closed this PR. Raised new PR against 0.10.0 branch. 
https://github.com/apache/kafka/pull/1687

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala (hunk 5 above)
Comment (omkreddy) [inline core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:411]: @hachikuji we are ignoring offset topic (non-matching unauthorized topic) 
testPatternSubscriptionWithTopicAndGroupRead test case. I can add separate test case, if needed.

**Pull request closed without merging**